src/libaktualizr/uptane/uptane_test.cc

```diff
@@ -227,13 +229,66 @@ TEST(Uptane, PutManifest) {
             "test-package");
 }
 
+class HttpPutManifestFail : public HttpFake {
+ public:
+  HttpPutManifestFail(const boost::filesystem::path &test_dir_in) : HttpFake(test_dir_in) {}
+  HttpResponse put(const std::string &url, const Json::Value &data) override {
+    (void)data;
+    return HttpResponse(url, 504, CURLE_OK, "");
+  }
+};
+
+int num_events_PutManifestError = 0;
+void process_events_PutManifestError(const std::shared_ptr<event::BaseEvent> &event) {
+  std::cout << event->variant << "\n";
+  if (event->variant == "PutManifestComplete") {
+    EXPECT_FALSE(std::static_pointer_cast<event::PutManifestComplete>(event)->success);
+    num_events_PutManifestError++;
+  }
+}
+
+/*
+ * Send PutManifestComplete event if send is unsuccessful
+ */
+TEST(Uptane, PutManifestError) {
+  TemporaryDirectory temp_dir;
+  auto http = std::make_shared<HttpPutManifestFail>(temp_dir.Path());
+
+  Config conf("tests/config/basic.toml");
+  conf.storage.path = temp_dir.Path();
+
+  auto storage = INvStorage::newStorage(conf.storage);
+  auto events_channel = std::make_shared<event::Channel>();
+  std::function<void(std::shared_ptr<event::BaseEvent> event)> f_cb = process_events_PutManifestError;
+  events_channel->connect(f_cb);
+  num_events_PutManifestError = 0;
+  auto sota_client = SotaUptaneClient::newTestClient(conf, storage, http, events_channel);
+  EXPECT_NO_THROW(sota_client->initialize());
+  auto result = sota_client->putManifest();
+  EXPECT_FALSE(result);
```

Comment:
Why do we expect this to fail? Is it just because of an invalid URL? It's worth making it explicit why this works as it does.

Reply:
This test checks if we get event 'PutManifestComplete' even when putManifest fails. It fails by returning non 200 return code.

I think the name of the test speaks for itself.

Reply:
It's true, the name does make it clear that the putManifest fails, but what is not immediately clear to me is WHY it fails.

Reply:
Ah, I missed that 504 when I was scanning the code before. My bad.